.circleci/dynamic_config.yml

```diff
@@ -0,0 +1,403 @@
+var_1: &cache_key v4-angular_devkit-14.17-{{ checksum "yarn.lock" }}
+var_1_win: &cache_key_win v4-angular_devkit-win-16.10-{{ checksum "yarn.lock" }}
+var_3: &default_nodeversion '14.17'
+var_4: &cache_paths
+  - node_modules
+  - packages/angular/cli/node_modules
+  - packages/angular/create/node_modules
+  - packages/angular/cwa/node_modules
+  - packages/angular_devkit/architect_cli/node_modules
+  - packages/angular_devkit/architect_cli/node_modules
+  - packages/angular_devkit/benchmark/node_modules
+  - packages/angular_devkit/build_angular/node_modules
+  - packages/angular_devkit/build_webpack/node_modules
+  - packages/angular_devkit/core/node_modules
+  - packages/angular_devkit/schematics/node_modules
+  - packages/angular_devkit/schematics_cli/node_modules
+  - packages/ngtools/webpack/node_modules
+  - packages/schematics/angular/node_modules
+  # Path where Arsenal Image Mounter files are downloaded.
+  # Must match path in .circleci/win-ram-disk.ps1
+  - ./aim
+
+# Filter to only release branches on a given job.
+var_5: &only_release_branches
+  filters:
+    branches:
+      only:
+        - main
+        - /\d+\.\d+\.x/
+
+# Executor Definitions
+# https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-executors
+executors:
+  action-executor:
+    parameters:
+      nodeversion:
+        type: string
+        default: *default_nodeversion
+    docker:
+      - image: cimg/node:<< parameters.nodeversion >>
+    working_directory: ~/ng
+    resource_class: small
+
+  test-executor:
+    parameters:
+      nodeversion:
+        type: string
+        default: *default_nodeversion
+    docker:
+      - image: cimg/node:<< parameters.nodeversion >>
+    working_directory: ~/ng
+    resource_class: large
+
+  windows-executor:
+    # Same as https://circleci.com/orbs/registry/orb/circleci/windows, but named.
+    working_directory: ~/ng
+    resource_class: windows.medium
+    shell: powershell.exe -ExecutionPolicy Bypass
+    machine:
+      # Contents of this image:
+      # https://circleci.com/docs/2.0/hello-world-windows/#software-pre-installed-in-the-windows-image
+      image: windows-server-2019-vs2019:stable
+
+# Command Definitions
+# https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-commands
+commands:
+  fail_fast:
+    steps:
+      - run:
+          name: 'Cancel workflow on fail'
+          when: on_fail
+          command: |
+            curl -X POST --header "Content-Type: application/json" "https://circleci.com/api/v2/workflow/${CIRCLE_WORKFLOW_ID}/cancel?circle-token=${CIRCLE_TOKEN}"
+
+  setup_windows:
+    steps:
+      - run: nvm install 16.10
+      - run: nvm use 16.10
+      - run: npm install -g yarn@1.22.10
+      - run: node --version
+      - run: yarn --version
+      - run: git config --global core.longpaths true
+
+  setup_linux:
+    steps:
+      - checkout
+      - run:
+          name: Rebase PR on target branch
```

Comment:
I see what you mean you good a point. Although I don't think this could ever happen in a PR, because when a commit is pushed the current running workflow becomes redundant and it is cancelled. 

I am not strongly about this, if you feel that we should do it once. I can definitely introduce a lighter version of `setup`.

Reply:
My concern here is that a push to the target branch has the possibility of a PR running early jobs against a different base commit than later jobs.
I think in this case, the consistency benefit should outweigh the test time for a lighter version of setup.